retrieve versions & meta with reverse dependencies

joseluis opened the pull request:
fixes: #8 

Made crate_reverse_dependencies() return a tuple of the three available API elements. Note: only the sync version really fetches the fields, while the async version uses a placeholder.

This is a working example. Save as `src/bin/new-rev-dep.rs` and run `cargo run` to check:
```rust
extern crate crates_io_api;
use crates_io_api::SyncClient;

fn main() {
    let client = SyncClient::new();
    let crate_name = "hyper-native-tls";
    let (revdep, revvers, meta) = client.crate_reverse_dependencies(crate_name).unwrap();

    println!("total={:?}", meta.total);
    println!("\ndependencies={:?}", revdep);
    println!("\nversions={:?}", revvers);
}
```

joseluis commented:
I first modified the `Dependencies` type to include the extra fields, but found out later while fixing #7 that it caused a conflict while fetching normal dependencies. So I found it clearer to create the `ReverseDependencies` type.

theduke commented on src/types.rs:
```diff
@@ -220,6 +222,13 @@ pub struct Dependencies {
 }
 
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct ReverseDependencies {
```
Since this is an internal type, it should only be `pub(crate)`

theduke commented on src/types.rs:
```diff
@@ -103,6 +103,8 @@ pub struct Version {
     pub license: Option<String>,
     pub readme_path: Option<String>,
     pub links: VersionLinks,
+    pub crate_size: Option<u64>,
```
Can you factor out the three new field additions into a separate commit?

joseluis replied:
Ok. I'm redoing this in a new branch, starting with this commit.

theduke commented on src/types.rs:
```diff
@@ -258,6 +267,8 @@ pub struct FullCrate {
     pub downloads: Downloads,
     pub owners: Vec<User>,
     pub reverse_dependencies: Vec<Dependency>,
+    pub reverse_dependencies_versions: Vec<Version>,
```
I assume the new response looks that way for caching purposes, but it's pretty confusing to use. I think the `version` and `de

We can make this nicer by adding a new type:

```rust
pub struct ReverseDependency {
  pub crate_version: Version,
  pub dependency: Dependency,
}
```

theduke replied:
adding an additional method that fetches a single page makes sense.

probably: 

```rust
fn crate_reverse_dependencies(&self, _crate: &str, page: u64, page_size: u64) ...
fn crate_reverse_dependencies_all(&self, _crate: &str) ...
```

joseluis replied:
I like that, I'm working on this.

theduke commented on src/types.rs:
```diff
@@ -258,6 +267,8 @@ pub struct FullCrate {
     pub owners: Vec<User>,
     pub reverse_dependencies: Vec<Dependency>,
+    pub reverse_dependencies_versions: Vec<Version>,
+    pub reverse_dependencies_meta: Meta,
```
that field is not really needed since we always fetch all dependencies anyway, so the `total` information would just by a `.len()` away.

theduke commented on src/async_client.rs:
```diff
@@ -218,10 +218,14 @@ impl Client {
         let owners = self.crate_owners(name);
         let reverse_dependencies = self.crate_reverse_dependencies(name);
+        // TODO:
+        // let (reverse_dependencies, reverse_dependencies_versions, reverse_dependencies_total) = self.crate_reverse_dependencies(name);
```
Do you need help in implementing this?

joseluis replied:
This should be fixed already. I'm sorry, I think I made a mess of the review process...

theduke commented:
Hey, thanks.

Added a review above.

joseluis commented:
@theduke I've redone the changes. This way is so much better.

Also, added the updated fields to a separate commit.

The only thing I've not finished is the async code, since it's not in use and I'm not very familiar with async. I just fixed the compiler errors and added a comment to fix it later.

joseluis commented:
I never used the github review before and I think I broke it when I force pushed the changes 😅

joseluis commented:
I'm gonna keep working on this. I've separated the adding new types into a different PR, and will fix the corresponding async code.

joseluis commented:
Finally I've fixed both the sync and async code.

I can't operate with the previous review, better close it and open another one if needed.

joseluis commented:
Hi, I believe the changes are solid but please don't hesitate to let me know if you'd prefer something done differently.

theduke commented:
Ah sorry for letting this drop, I'll review now.

theduke commented on src/types.rs:
```diff
@@ -223,6 +223,27 @@ pub struct Dependencies {
+
+// This is how reverse dependencies are received
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct ReverseDependenciesAsReceived {
```
This should not be public.
It can be `pub(super)` instead.

joseluis replied:
Much better this way indeed. I didn't know that was an option.

theduke commented on src/async_client.rs:
```diff
@@ -93,36 +93,43 @@ impl Client {
+            c.get::<ReverseDependenciesAsReceived>(&url).and_then(move |rdeps|
+                -> Box<Future<Item = ReverseDependencies, Error = Error> + Send> {
+
+                for d in rdeps.dependencies.iter() {
```
This code is duplicated twice, can you put in on a crate-private method on `ReverseDependencies`?

Somethink like `fn from_received(...)..`

joseluis replied:
Done.

theduke commented on src/async_client.rs:
```diff
@@ -93,36 +93,43 @@ impl Client {
+            )).unwrap();
+
+            c.get::<ReverseDependenciesAsReceived>(&url).and_then(move |rdeps|
+                -> Box<Future<Item = ReverseDependencies, Error = Error> + Send> {
```
nit: this should be `dyn Future<...>`

joseluis replied:
Ok. I don't think I'm really understanding this, but I just added the `dyn` keyword.

theduke commented:
A few small changes, then good to merge.

joseluis commented:
All requested changes done. Test result: ok.
